pkg/generators/openapi.go

```diff
@@ -662,3 +674,95 @@ func (g openAPITypeWriter) generateSliceProperty(t *types.Type) error {
 	g.Do("},\n},\n},\n", nil)
 	return nil
 }
+
+// CustomArgs is used by the gengo framework to pass args specific to the API linter
+type CustomArgs struct {
+	ReportFilename string
+}
+
+type apiLinter struct {
+	rules          []APIRule
+	violations     []apiViolation
+	reportFilename string
+}
+
+func newAPILinter(reportFilename string) *apiLinter {
+	return &apiLinter{
+		rules: []APIRule{
+			&rules.NamesMatch{},
+		},
+		reportFilename: reportFilename,
+	}
+}
+
+type apiViolation struct {
+	// Name of rule from APIRule.Name()
+	rule string
+
+	packageName string
+	typeName    string
+
+	// Optional: name of field that violates API rule. Empty fieldName implies that
+	// the entire type violates the rule.
+	field string
+}
+
+// APIRule is the interface for validating API rule on Go types
+type APIRule interface {
+	// Validate evaluates API rule on type t and returns a list of field names in
+	// the type that violate the rule. Empty field name [""] implies the entire
+	// type violates the rule.
+	Validate(t *types.Type) ([]string, error)
+
+	// Name returns the name of APIRule. The name should not contain ','.
+	Name() string
+}
+
+func (l *apiLinter) validate(t *types.Type) error {
+	for _, r := range l.rules {
+		glog.V(5).Infof("validating API rule %v for type %v", r.Name(), t)
+		fields, err := r.Validate(t)
+		if err != nil {
+			return err
+		}
+		for _, field := range fields {
+			l.violations = append(l.violations, apiViolation{
+				rule:        r.Name(),
+				packageName: t.Name.Package,
+				typeName:    t.Name.Name,
+				field:       field,
+			})
+		}
+	}
+	return nil
+}
+
+func (g *openAPIGen) Finalize(c *generator.Context, w io.Writer) error {
+	// If report file isn't specified, log and return error
+	if len(g.linter.reportFilename) == 0 {
+		for _, v := range g.linter.violations {
+			glog.Errorf("API rule violation: %s,%s,%s,%s", v.rule, v.packageName, v.typeName, v.field)
+		}
+		if len(g.linter.violations) > 0 {
+			return fmt.Errorf("API rule violations exist")
+		}
+		return nil
+	}
+	// Otherwise, print the error to report file and return nil
+	f, err := os.Create(g.linter.reportFilename)
```

Comment:
check for "-" as the output, and if so, write to os.Stdout. Also update docs for the flag.

Reply:
Speaking of the flag, is it actually wired up? I don't see it.

Reply:
> Speaking of the flag, is it actually wired up?

Not yet in this PR, I experimented in https://github.com/roycaihw/kubernetes/commit/2892801c5dfc14ecf4e389e56151f8e3801298df with the old main.go